packages/application-shell/src/components/navbar/navbar-new/navbar.tsx

```diff
@@ -111,29 +112,32 @@ export const ApplicationMenu = (props: ApplicationMenuProps) => {
   const menuItemTop = menuItemBoundingClientRect?.top || 0;
   const menuItemBottom = menuItemBoundingClientRect?.bottom || 0;
 
-  const callbackFn: IntersectionObserverCallback = (entries) => {
-    const [entry] = entries;
-    const doesSubmenuFitViewportBelowMenuItem =
-      entry.boundingClientRect.height +
-        (props.isMenuOpen ? menuItemTop : menuItemBottom) >
-      window.innerHeight;
-    // if the submenu does not fit at the bottom of the viewport (below the menu item)
-    if (doesSubmenuFitViewportBelowMenuItem) {
-      setIsSubmenuAboveMenuItem(true);
-      setSubmenuVerticalPosition(
-        window.innerHeight - (props.isMenuOpen ? menuItemBottom : menuItemTop)
-      );
-      // show the submenu above the menu item
-    } else {
-      setIsSubmenuAboveMenuItem(false);
-      setSubmenuVerticalPosition(
-        props.isMenuOpen ? menuItemTop : menuItemBottom
-      );
-    }
-  };
+  const callbackFn: IntersectionObserverCallback = useCallback(
+    (entries) => {
+      const [entry] = entries;
+      const doesSubmenuFitWithinViewportBelowMenuItem =
+        entry.boundingClientRect.height +
+          (props.isMenuOpen ? menuItemTop : menuItemBottom) >
+        window.innerHeight;
+      // if the submenu does not fit at the bottom of the viewport (below the menu item)
+      if (doesSubmenuFitWithinViewportBelowMenuItem) {
+        setIsSubmenuAboveMenuItem(true);
+        setSubmenuVerticalPosition(
+          window.innerHeight - (props.isMenuOpen ? menuItemBottom : menuItemTop)
+        );
+        // show the submenu above the menu item
+      } else {
+        setIsSubmenuAboveMenuItem(false);
+        setSubmenuVerticalPosition(
+          props.isMenuOpen ? menuItemTop : menuItemBottom
+        );
+      }
+    },
+    [menuItemBottom, menuItemTop, props.isMenuOpen]
+  );
   useLayoutEffect(() => {
     const observer = new IntersectionObserver(callbackFn, {
```

Comment:
Is it expected to create a new observer instance every time the layout effect runs? 🤔

Looking at our implementation of `useMutationObserver`, we cache the instance to not create it every time: https://github.com/commercetools/ui-kit/blob/73bc8841022ea3b46dc9e2b75d3f9fd3e1d4fff6/packages/hooks/src/use-mutation-observer/use-mutation-observer.ts#L68-L71

Reply:
I don't think so. I guess that would be a nice enhancement 🙂
I just wanted to double check if your expectation here is to:
- create our own `useIntersectionObserver` hook implementation (based on the implementation of our `useMutationObserver`)
- use [this](https://github.com/jaredLunde/react-hook/tree/master/packages/intersection-observer#readme) polyfill
- simply create `IntersectionObserver` cache without any additional hooks?

Reply:
Just the caching part. You can also use a `ref` to cache it.

Reply:
It also depends on what are you observing. Is it a static element (you then need only 1 observer) or multiple elements, in which case you might need multiple instantiations.

Reply:
Looking at the implementation, it seems that the intersection observer is needed for the `<MenuGroup>` component. Would it make sense to define the observer in that component instead of the parent `<ApplicationMenu>`? Then you have one observer for each rendered `<MenuGroup>`.

Reply:
Thanks.

> Would it make sense to define the observer in that component instead of the parent `<ApplicationMenu>`

Yes, it's definitely [a potential improvement](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API#:~:text=We%20could%20opt%20to%20monitor%20multiple%20elements%20for%20visibility%20intersection%20changes%20with%20respect%20to%20the%20viewport%20by%20calling%20observer.observe()%20for%20each%20of%20those%20elements%2C%20if%20we%20wanted%20to%20do%20so.), but for now I wanted to avoid dynamically creating all the refs and prop-drilling them. I have some experience with something similar in the LocalizedRichTextEditor, but I'm a bit concerned about our time-frame for this feature. But I can look into that.

Reply:
Thanks. It's fine for me then as long as we are aware of potential implications.

Reply:
Perhaps we could add that to our technical depth list for after completing phase 1? As long as it's not a critical issue.

Reply:
We could also just merge this branch for now since it includes a lot of other changes and then take a closer look at this separately.

Reply:
Yes please go ahead, don't feel blocked by this by any means. I was mostly curious about understanding the background a bit more.